mindmeld/system_entity_recognizer.py

```diff
@@ -41,61 +44,101 @@ class SystemEntityError(Exception):
+    """
+
+    @staticmethod
+    def get_instance(app_path):
+        """ Static access method.
+        In general we want to find the system entity recognizer from the application config.
+        If the application configuration is empty, we do not use Duckling.
+        Otherwise, we return the Duckling recognizer with the URL defined in the application's
+          config, default to the DEFAULT_DUCKLING_URL.
+
+        Args:
+            app_path: The application path
+
+        Returns:
+            (SystemEntityRecognizer): A SystemEntityRecognizer instance
+        """
+        if is_duckling_configured(app_path):
+            url = get_system_entity_url_config(app_path=app_path)
+            return DucklingRecognizer.get_instance(url)
+        else:
+            return DummySystemEntityRecognizer.get_instance()
+
+    @abstractmethod
+    def get_response(self, data):
+        pass
+
 
-    TODO: Abstract this class into an interface and implement the duckling
-    service as one such service.
+class DummySystemEntityRecognizer(SystemEntityRecognizer):
+    """
+    This is a dummy recognizer which returns empty list and NO_RESPONSE_CODE.
     """
 
     _instance = None
 
-    def __init__(self, app_path=None):
+    def __init__(self):
+        if not DummySystemEntityRecognizer._instance:
+            DummySystemEntityRecognizer._instance = self
+        else:
+            raise Exception("DummySystemEntityRecognizer is a singleton")
+
+    @staticmethod
+    def get_instance():
+        if not DummySystemEntityRecognizer._instance:
+            DummySystemEntityRecognizer()
+
+        return DummySystemEntityRecognizer._instance
+
+    def get_response(self, data):
+        del data
+        return [], NO_RESPONSE_CODE
+
+
+class DucklingRecognizer(SystemEntityRecognizer):
+    _instances = {}
+
+    def __init__(self, url=None):
         """Private constructor for SystemEntityRecognizer. Do not directly
-        construct the SystemEntityRecognizer object. Instead, use the
+        construct the DucklingRecognizer object. Instead, use the
         static get_instance method.
 
         Args:
-            app_path (str): A application path
+            url (str): Duckling URL
         """
-        if SystemEntityRecognizer._instance:
-            raise Exception("SystemEntityRecognizer is a singleton")
-        else:
-            if not app_path:
-                # The service is turned on by default
-                self._use_duckling_api = True
-            else:
-                self._use_duckling_api = is_duckling_configured(app_path)
+        if url in DucklingRecognizer._instances:
+            raise Exception("DucklingRecognizer is a singleton")
 
-        self.app_path = app_path
-        SystemEntityRecognizer._instance = self
+        self.url = url
+        DucklingRecognizer._instances[url] = self
 
     @staticmethod
-    def get_instance(app_path=None):
+    def get_instance(url=None):
```

Comment:
Assign the default here to `DEFAULT_DUCKLING_URL` and you don't have to do the conditional assignment below.

Reply:
The caller of this function can still pass None into get_instance, so I thought it is probably neat to keep the conditional assignment below.

Reply:
What does it mean for the caller to pass None when it just uses the `DEFAULT_DUCKLING_URL ` in that case?

Reply:
The caller might read the URL from some config sources which return None